### PR TITLE
Implement `Default` trait for structs generated from `@oneOf`

### LINKED ISSUE #18: Support returning `None` as default for unused union types
Related to https://github.com/Shopify/fulfillment-constraints-sample-function-app/pull/2#discussion_r1072815329.

We've yet to find a solution or workaround to being forced to return `None` for all unused union fields when returning operations in the `fulfillment_constraints` API. Currently we support two constraints, but i the future we intend to support many more, so we don't want to be forced to pass `None` for a large list of constraints when we only want to return one.

For example, currently we have to use:
```rust
output::Operation {
            must_fulfill_from_same_location: Some(output::MustFulfillFromSameLocation { cart_line_ids }),
            must_fulfill_from: None,
            some_other_constraint: None,
}
```

What we'd like to use:

```rust
output::Operation {
            must_fulfill_from_same_location: Some(output::MustFulfillFromSameLocation { cart_line_ids }),
}
```

### PR DESCRIPTION
**Pull request by surma**
Closes #18

### CODE REVIEW

File: shopify_function_macro/src/lib.rs [hunk 1]
```diff
@@ -127,14 +128,73 @@ pub fn generate_types(attr: proc_macro::TokenStream) -> proc_macro::TokenStream
...
+                == name.as_ref()
+    } else {
+        false
+    }
+}
+
+#[proc_macro_derive(OneOfDefault)]
+pub fn one_of_default_derive(item: proc_macro::TokenStream) -> proc_macro::TokenStream {
+    let parsed_item: syn::Item = parse_macro_input!(item);
+    let syn::Item::Struct(parsed_struct) = parsed_item else { return quote!{}.into()};
```
Comment (adampetro) [inline shopify_function_macro/src/lib.rs:160]: TIL about `let-else`

File: shopify_function_macro/src/lib.rs [hunk 2]
```diff
@@ -127,14 +128,73 @@ pub fn generate_types(attr: proc_macro::TokenStream) -> proc_macro::TokenStream
...
+    // If every field in the struct is an `Option`, it's the result
+    // of a `@oneOf` directive.
+    let is_oneof_struct = parsed_struct
+        .fields
+        .iter()
+        .all(|field| is_type(&field.ty, "Option"));
```
Comment (adampetro) [inline shopify_function_macro/src/lib.rs:166]: Not necessarily true, I believe? (sorry if I'm being pedantic) But nonetheless it does mean that we can derive `Default`

File: shopify_function_macro/src/lib.rs [hunk 3]
```diff
@@ -127,14 +128,73 @@ pub fn generate_types(attr: proc_macro::TokenStream) -> proc_macro::TokenStream
...
+    if let syn::Type::Path(path) = ty {
+        path.path.segments.len() == 1
+            && path
+                .path
+                .segments
+                .first()
+                .unwrap()
+                .ident
+                .to_string()
+                .as_str()
```
Comment (adampetro) [inline shopify_function_macro/src/lib.rs:150]: I think we can make use of [this `PartialEq` impl](https://docs.rs/syn/latest/syn/struct.Ident.html#impl-PartialEq%3CT%3E-for-Ident):
```suggestion
```

**Review by adampetro — APPROVED** [review]

**Comment by surma** [comment]
Closing in favor of upstream fix.

**Pull request closed without merging**